### ISSUE TITLE
MongoDB Connector

### ISSUE THREAD

**Issue by marktani** [prisma/prisma#1643]
This feature requests serves as a central place to discuss development and progress for the MongoDB connector. Please contribute in the comments below. I will update the top comment as consensus forms.

**Comment by lolcoolkat** [comment]
Can DynamoDB support be added pls! :D

**Comment by ForsakenHarmony** [comment]
Make a new issue for it

**Comment by ookangzheng** [comment]
Mysql pls

**Comment by alexandrakl** [comment]
I would like to work on this! Where is a good place to start looking? How to wrap head around the task and break it into its constituent parts? Please provide some guidelines if possible, would be much appreciated! Thanks for your work on this project, it is very cool, reminds me of RESTHeart :)

**Comment by cbioley** [comment]
@ookangzheng Have a look at the project README file. It says "Prisma can be used for MySQL Databases out of the box" ;)

**Comment by sorenbs** [comment]
Hi @alexandrakl it would be great to collaborate on this!

The first step is to map out the kind of queries that are particularly well supported by Mongo as well as queries currently supported by Prisma that does not fit the document model.

Based on this we can design the API for Document-flavoured GraphQL.

Feel free to dump your thoughts in this thread so we can get the conversation going :-)

**Comment by GilUpstream** [comment]
@alexandrakl @sorenbs I'd love to work with you all on this.

**Comment by sorenbs** [comment]
Here's an outline of some things we need to consider:

* align with existing Mongo-based GraphQL APIs (vulcanJs)
* active/passive (does schema migration even make sense?)
* mutations (can we support the powerful nested mutations? relations?)
* queries (What are the query limitations?)
* subscriptions

**Comment by sorenbs** [comment]
## Top level fields

For each type with an `id` field, the following fields are generated:

> Types without an `id` field are called `embedded types` and are stored as nested documents.

> The `id` field is represented as `_id` in the mongo document per normal mongo convention. Prisma does not support having both `_id` and `id` fields.

> The type `User` is used as an example throughout this document

### Query

- `user` returns a single user
- `users` returns a list of users
- `usersConnection` returns a relay style Connection of users

> In addition to the generated query fields, the following field is present:

- `node` returns a single node by id

#### field name generation

The `user` field name is generated by taking the type name and lowercasing the first letter

The `users` and `usersConnection` field names are generated by taking the type name and lowercasing the first letter and then using the [evo-inflector library](https://github.com/atteo/evo-inflector)

### Mutation

- `createUser` create a single user
- `updateUser` update a single user
- `deleteUser` delete a single user
- `upsertUser` update a single user if exists, otherwise create
- `updateUsers` update multiple users
- `deleteUsers` delete multiple users

> Note: The API has no explicit mutations for managing relations. Instead relations are managed through nested mutations #1280 

> MongoDB has no concept of foreign key constraints and cascading deletes. Prisma ensures that relations stays valid and that cascading deletes are enforced.

Examples:

```
A -> B
```

- Deleting `B` will fail if the relation is required and there is no cascading delete
- Deleting `B` will succeed if the relation is required and there is cascading delete
- Deleting `B` will succeed if the relation is not required and there is no cascading delete. In this case the relation field will be set to null
- Deleting `B` will (obviously) succeed if the relation is not required and there is cascading delete 

> Note: MongoDB has no concept of transactions that span more than one document. As mutations that involve relations or mutate many nodes, involve multiple documents, it is possible that the mutation will be only partially applied if the server is stopped before the mutation completes.

## Details

```graphql
type Query {
  user(where: UserWhereUniqueInput): User

  users(
    where: UserWhereInput
    orderBy: UserOrderByInput
    skip: Int
    before: String
    after: String
    first: Int
    last: Int
  ): [User!]!

  usersConnection(
    where: UserWhereInput
    orderBy: UserOrderByInput
    skip: Int
    before: String
    after: String
    first: Int
    last: Int
  ): UserConnection!
}

type Mutation {
  createUser(data: UserCreateInput!): User!
  updateUser(data: UserUpdateInput!, where: UserWhereUniqueInput!): User
  deleteUser(where: UserWhereUniqueInput!): User
  upsertUser(where: UserWhereUniqueInput!, create: UserCreateInput!, update: UserUpdateInput!): User

  updateUsers(data: UserUpdateInput!, where: UserWhereInput!): BatchPayload
  deleteUsers(where: UserWhereInput!): BatchPayload
}
```

## Nested Mutations

Mongo supports arbitrarily nested documents but has no understanding of relations between documents.

Prisma supports two ways to model related data:

- relations
- embedded types.

`relations` is a great fit for relational databases and Prisma implements embedded types for SQL databases by automatically managing all the tables and foreign keys required.

In the same way, `embedded types` is a great fit for mongos document model. Prisma implements relations by performing multiple inserts/updates or queries under the hood.

### When is a type embedded?

In Prisma a type is considered to be embedded when it has no `id` field. Types without an `id` field can only ever exist when referenced from an "owning" type. The mongo connector takes advantage of this fact by using mongos native capability to nest documents. This can result in improved performance compared to SQL.

> Note: MongoDB has a storage limit of 16mb per document. This limits the aggregate size of a single node with embedded types. The SQL connector does not have this limit.

### handling of relations

> todo: specify how mutations and queries are handled for embedded types

### scalar lists

> todo: specify how scalar lists are implemented. They should be embedded. Document how each operation maps to mongos primitives. Document storage limit

**Comment by juicycleff** [comment]
There needs the be a common api or interface so one can switch to different connectors.

**Comment by sorenbs** [comment]
That's a great point @juicycleff 

We plan to support database specific APIs as well a a Core API. Most of the database specific APIs will have significant overlap with the core API making it easy to switch between connectors. For Mongo and SQL specifically we will be able to have almost 100% overlap :-)

**Comment by Yogu** [comment]
If you need some inspiration for a GraphQL type system of a document database, have a look at [cruddl](https://github.com/AEB-labs/cruddl). This is a TypeScript library that servers a GraphQL API similar the one of prisma, but it uses ArangoDB as database (might also be interesting for #1645) is built to embrace the possibilities of document databasees like embedding. We invented [four kinds of object types](https://github.com/AEB-labs/cruddl/blob/master/docs/modelling.md) to support different use cases of embedding and relating objects - like when id should be auto-generated and whether omitted fields in update mutations should remove them or leave them unchanged.

**Comment by DavyBello** [comment]
Have a look at [graphql-compose](https://github.com/graphql-compose/graphql-compose) it's a toolkit for construction of complex GraphQL schemas. Provides bunch of plugins for type generation. Allows to build your own plugins/helper-functions for type construction and modification and [graphql-compose-mongoose](https://github.com/graphql-compose/graphql-compose-mongoose) a Mongoose model converter to GraphQL types with resolvers. They also have type generator plugins for elasticsearch and aws

**Comment by duytai** [comment]
@sorenbs in my opinion, we should consider a `model` without `embedded` type to remove the complexity of mongodb connector. As far as I know: 
1. Current generated Prisma api does not support `embedded` type
2. We just implement connector to support current Prisma API

**Comment by sorenbs** [comment]
@duytai i agree. 

The term embedded type is used here only to describe a certain way to lay out your data. Prismas SQL connector has no understanding of that term and neither will the Mongo connector. The final api will be almost identical apart from less powerful filters.

**Comment by DakshMiglani** [comment]
any updates on this?

**Comment by ojongerius** [comment]
Is someone working on this at the moment?

**Comment by marktani** [comment]
Yes, we are 🙂

We plan to start a preview for the MongoDB connector later this month. Please reach out to me in [Slack](https://slack.graph.cool) (@nilan) if you're interested in participating.

**Comment by lfades** [comment]
An update for this note:

> Note: MongoDB has no concept of transactions that span more than one document. As mutations that involve relations or mutate many nodes, involve multiple documents, it is possible that the mutation will be only partially applied if the server is stopped before the mutation completes.

Transactions will be supported in [4.0](https://www.mongodb.com/transactions)

**Comment by lfades** [comment]
> todo: specify how mutations and queries are handled for embedded types

In Mongodb joins/relations are not recommended for cases where the related documents don't change often, better said, we have a lot more of queries than mutations, for example, suppose that our `User` has a `books` fields which can be a one-to-many relation or just an embedded field with the books instead, A `Book` will probably never change so we don't need a one-to-many relation here.

I believe embedded types can be handled in the same way Prisma handles relations, but every embedded document should contain an `id` (and the operations for update/remove/insert will be using different mongodb operators). In most cases embedded documents will contain an `id`, or something that will allow them to be identified inside the array (like the `index`), if the embedded document is just an object and not an array, all operations will be for that document.

**Comment by nhuesmann** [comment]
@sorenbs Have you considered including any mongoose-specific abstractions like [virtuals](http://mongoosejs.com/docs/guide.html#virtuals) and references/reference [population](http://mongoosejs.com/docs/populate.html) (i.e. using [references](https://docs.mongodb.com/manual/tutorial/model-referenced-one-to-many-relationships-between-documents/) as an alternative to embedded documents)? I've really liked utilizing mongoose's populate operator to keep document sizes to a predictable limit, avoid arrays growing without bound, and normalize the data as much as possible. Virtuals/getters and setters are more of a nice-to-have but have still proven to be handy in certain situations, acting like Redux selectors.

I know the purpose of the mongodb connector isn't to be a mongoosejs connector, however I anticipate a large number of developers having likely interacted with mongoosejs as much if not more than the mongodb API directly, so having some familiar features might make the adoption process easier.

Also, do you plan to support [GeoJSON](https://docs.mongodb.com/manual/reference/geojson/index.html) object types for geospatial queries? They've been indispensable in a few projects and I'd really hope for them to be supported.

**Comment by heysailor** [comment]
+1 for GeoJSON. My primary use case for mongo is to enable efficient geospatial queries. These complex queries can be done with mongo directly, but the ability to resolve GeoJSON data through Prisma would be very handy.

**Comment by debianw** [comment]
hi @marktani. Do you have this MongoDB connector in a specific branch, what's the name of the branch. I would like to take a look to the code. Thanks

**Comment by heysailor** [comment]
...just to clarify my earlier comment re GeoJSON, the key to allowing location queries is a location field on the document. It'd be great if the Prisma SDL included a type such as `GeoPoint` (as it does other types, like `DateTime`). Then a mongo geolocation query could be done independent of Prisma, or even awesomer, through Prisma with its paging support.

**Comment by ArvinGanesan** [comment]
@alexandrakl,@sorenbs  - I would love to contribute to this project. I hope that it's not too late. Currently, I'm studying data science at Berkeley.  And I have been a fan of this project. Full Disclosure - I only just got started with Scala. I would love for this to be a learning experience.

**Comment by sorenbs** [comment]
@ArvinGanesan - it is never too late to contribute :-)

I think the MongoDB connector would be too big of a mouthful for you to tackle if you are new to Scala. As a first step I would recommend that you clone the repo and get to a point where you can run the tests locally.

Issues tagged `good-first-issue` https://github.com/graphcool/prisma/issues?q=is%3Aopen+is%3Aissue+label%3Agood-first-issue might be a good place to start. Alternatively you could find a reported issue that you find interesting and create a test case that demonstrates the issue. This will help us prioritise and fix that issue and it helps you get an understanding for the code base.

Please keep in mind that the code base is moving fast and we might not be able to merge your PR.

@debianw We will begin implementing the MongoDB connector in a few weeks and will link the branch from this issue so you can follow along. Our initial plan was to begin the implementation last month, but we had to rearrange some work tasks.

@heysailor - thanks for your input on GeoJSON. This is definitely on our radar!

**Comment by ArvinGanesan** [comment]
@sorenbs - I ran in to some sbt build error. I did create an issue for that - sbt build error #2349

Any help is appreciated!

**Comment by jay-jlm** [comment]
I  this being developed in a public branch somewhere? 
I'm curious how it's progressing since the initial delivery timeframe was back in April.
We have a new project rapidly approaching its prototype phase and we are asking ourselves  if we can count on this being public by then.

**Comment by sorenbs** [comment]
@jay-jlm We are planning to release a public roadmap soon. Although our planning is not final yet, you should expect a beta release of the MongoDB connector in August.

**Comment by bazaglia** [comment]
@sorenbs I read in prisma blog that version 1.11 (next one) will include the beta support for mongodb. Does this roadmap still the same?

**Comment by sorenbs** [comment]
@bazaglia that is inaccurate - sorry about the confusion. August is the current plan :-)

**Comment by dandv** [comment]
> Transactions will be supported in 4.0

MongoDB 4.0 was released on June 28.

https://www.mongodb.com/transactions

**Comment by nolandg** [comment]
https://www.prisma.io/ lists MongoDB as a selling feature but this thread seems to contradict that...am I missing something? is there another branch I can checkout?

**Comment by lfades** [comment]
@nolandg The mongodb connector is probably the closest connector coming out soon, it will be released in August

**Comment by gianpaj** [comment]
happy 10th of August 😎 🏖

**Comment by do4gr** [comment]
The work for the Mongo connector is underway and there is a related discussion about embedded types. So if someone has strong opinions around embedded types you can chime in here: https://github.com/prisma/prisma/issues/2836

**Comment by jay-jlm** [comment]
Happy 24th of August, evryone!
Say what you want about the Prisma team, but you can't deny they know how to keep the graphql community on their toes!  😆😆

**Comment by firrae** [comment]
Are there any updated timelines as we're about to start the first week of September?

**Comment by do4gr** [comment]
We are actively working on it. You can follow this PR https://github.com/prisma/prisma/pull/2883

**Comment by ysantalla** [comment]
How can I test the new functionalities???

**Comment by do4gr** [comment]
Once we have a public preview available we'll report back here.

**Comment by do4gr** [comment]
We now have a very early prototype available for testing. We decided to implement the new feature of [embedded types](https://github.com/prisma/prisma/issues/2836) first to get feedback on that early on. You can head over [here](https://github.com/prisma/Mongo-Connector-Preview/blob/master/README.md) to check out how to try out the early preview.

**Comment by iamrommel** [comment]
The website says it working on mongodb but you comments says beta, prototype, in-progress, I highly suggesst better removed that "check" on mongodb because it's totally misleading. 

Admit it, you just want a marketing bait!!!!!

Grow up prisma and separate yourself from those poor documented and poor company that just want attention based on non-existing products. I've been to that, I've been to the company that just sells "photoshoped" products, products that exists only on photoshop, but no really running product.

Grow up and fixed yourself!!!

**Comment by marktani** [comment]
Thanks for sharing your point of view, @iamrommel! 🙂You are expressing what probably many are thinking. I can feel your frustration! I am sorry to hear that you feel misled, and I offer a different perspective.

It is not our intention to use marketing terms or click bait titles to draw in more attention. In fact, we are positively overwhelmed about the attention and the passionate feedback for the MongoDB connector, and many other connectors! It makes me smile everytime there's a new feature request for a new DB connector, or someone offering their help in pushing a particular connector forward.

Driven by our excitement and the community feedback, we are working hard on building more connectors and are really excited about its potential. We have validated the MongoDB connector in thorough internal tests and many discussions and demos with customers. Hence, we are convinced that the MongoDB connector will provide a lot of value, and we want to share our excitement and conviction with the community on our homepage!

However, we don't always find the right ways and words to express this excitement, what we think about a feature, or what state it's currently in. And we have been working much longer on the MongoDB connector than anticipated. These are aspects where we can improve a lot, and I really appreciate your participation by reminding us of this.

One particular step of improvement that I can think of is to add a "beta" label whereever the MongoDB connector is mentioned on our homepage. I will bring this up to the team.

I hope you will try out the preview soon, and I'm curious to hear what you think about it. You can always find me in Slack (@nilan), if you have any questions or want to share anything else. Thanks 🙌

**Comment by nolandg** [comment]
Great answer. I think everyone needs to realize that making money off of open source involves walking some tricky and thin lines and it will be impossible to do it all perfectly and keep everyone happy. This must be even more difficult when you're spending someone else's $4.5M and they need to see returns.  I accept commercial open source being a bit trigger happy in the marketing department as a necessity of getting this great software. Cheers to Prisma, you're making me money.

**Comment by iamrommel** [comment]
@nolandg i know everyone needs a move to make some money. But i guess everyone needs be careful too. 


My point on my post is being true on prisma for telling user/customer about thier product status. The mongodb CRUD from the website has been check maybe 2 or 3months (or even more) but there is no valid resources for it, i've been checking on it for them several times but none was provided.

The information is misleading that is what want to emphasize.

I now remember Aesop's fable about the boy and the wolf. The boy always telling lies about the attack of the wolf, and in the end when real wolf attacks, no one listen. So of prisma continues to feed us wrong/misleading/click bait thing, no one would listen even the real feature was realease and then lowering the confidence to the company too.

I'm not throwing stones at prisma, i'm just giving harsh feedbacks. And that's what REAL friends do, giving harsh comments and not sweet compliments. :)

**Comment by bmayen** [comment]
That's an interesting criteria for friendship 🤔
Regardless, this is a professional setting, not a friendship, and real professionals conduct themselves accordingly. Your tone here is counterproductive and overshadows any valid points your argument might have. Shouting "grow up" at the devs is a bit ironic, no?

**Comment by lfades** [comment]
@iamrommel We understand your point, Prisma still doesn't has the mongodb connector, I even remember having a preview of the Beta of the new Prisma that was going to replace Graphcool and asking why Mongo was there, they had plans to have the connector released very soon, it didn't happen but that doesn't means we need to be rude about it, it's open source and it's free, they surely have more priorities and we can always be nice about this stuff, I agree with your point, but not with "harsh comments and not sweet compliments".

**Comment by iamrommel** [comment]
![image](https://user-images.githubusercontent.com/177265/45657519-4ae88c00-bb1d-11e8-8c8f-1a42e6a4fa2f.png)

Maybe i'm on the wrong neighborhood, because "constructive feedback" are so 1900s, and  harsh comments are future (or maybe i'm living in the future.)

If anyone is offended, I sincerely apologize!

**Comment by jay-jlm** [comment]
@iamrommel  ouch, it must be indeed very self-reassuring to think that one's ideas come from the future. 
Sadly, I'm afraid you just happen to lack basic understanding of how basic economics work.

Companies must able to market future technologies because you need to build visibility for your product before it reaches the market so that demand has scaled to a sustainable level by the time does. 

Let me offer a different perspective than yours about "growing up": being a "grown-up" (developer) is all about relying on yourself to evaluate which and when to adopt and up-and-coming technology,  and not about expecting Prisma's mkt department to do this job for you.

**Comment by iamrommel** [comment]
@jay-jlm, 
Companies must able to market future technologies (that they have already committed resources to develop) because you need to build visibility for your product before it reaches the market so that demand has scaled to a sustainable level by the time does.

These are the things that I hate these days, as i have said, i've been to companies like these, just selling ideas from "photoshopped" products. It's very hard for the customer to accept that they wanted the product now and it's not yet available. You tell your customer that your prodcut are ready but when they want it now, they tell you back it under massive developement. They marketting specifically tells you that it's ready to roll and when you are going to use it, they'll drop you a bomb that it's not ready.

Remember my point about is about truth about the status. Mongodb CRUD feature is there for a couple of months but haven't seen any actual implememtation. If only that is stated clearly (coming soon, beta) then i'm pacified. And that is how to tell your customers about the future ideas (telling the thruth that it is beta or coming soon)

**Comment by marktani** [comment]
Even though there have been a lot of messages since my last post, I don't think I can add anything of value to the conversation at this point which I haven't said already. I think everyone had the chance to share their input 🙂

This thread is not right the place to continue this discussion. I would encourage everyone to start a new discussion in the feedback section of our Forum if there is further need to talk: https://www.prisma.io/forum/c/feedback. I appreciate any and all input there.

If you are interested in the MongoDB preview, head over to https://github.com/prisma/Mongo-Connector-Preview.

And again, you can always find me in Slack (@nilan) to talk about anything.

Thank you.

**Comment by do4gr** [comment]
We just merged Join-Relations (relations between models in different collections) into the Mongo Connector preview. If you want to try it out, I’ve updated the preview page. As always: feedback is very welcome. https://github.com/prisma/Mongo-Connector-Preview/blob/master/README.md (ed

**Comment by do4gr** [comment]
This is a proposal for the directives to  be introduced for the Mongo connector. 

We are already using connector specific directives in the Postgres Passive connector. https://www.prisma.io/docs/data-model-and-migrations/introspection-mapping-to-existing-db-soi1/#overview

|      Postgres Directive      | Active | Passive |                                                               |
|:----------------------------:|--------|---------|---------------------------------------------------------------|
|  @pgTable(name: "X")         |   opt  |   opt   | can be provided to rename tables                              |
| @pgColumn(column: "X")       |   opt  |   opt   | can be provided to rename columns                             |
| @pgRelation(column: "X")     |   opt  |   req   | has to be provided to identify fields with relation ids       |
| @pgRelationTable(table: "X") |    -   |   req   | identifies a relation table                                   |
|                              |        |         |                                                               |
|        Mongo Directive       |        |         |                                                               |
| @mongoCollection(name: "X")  |   opt  |   opt   | can be provided to rename collections                         |
| @mongoField(field:"X")       |   opt  |   opt   | can be provided to rename fields                              |
| @mongoRelation               |   req  |   req   | explicitly determine the side that stores inline relation ids |
| @mongoRelation(field:"X")    |   opt  |   opt   | can be provided to rename inline relations and embedded types |




# Active Connector

Bi-Directional Join Relation

```graphql
type Parent{
    name: String
    child: Child @mongoRelation    # required! this side has inline ids, autogenerated field names
}

type Child{
    name: String
    parent: Parent
}
```

Uni-Directional Join Relation

```graphql
type Parent{
    name: String
    child: Child     # optional. no directive necessary since there is only one relationField
}

type Child{
    name: String
 }
```

Embedded Type Relation

```graphql
type Parent{
    name: String
    child: Child     # optional. embedded under fieldname
}

type Child @embedded {
    name: String
}
```

Embedded Type To Other Non-Embedded Type Relation

```graphql
type Parent{
    name: String
    child: Child     # optional. stored under fieldname
}

type Friend{
    name: String
}

type Child @embedded {
    name: String
    friend: Friend  # optional. stored under fieldname
}
```

# Passive Connector

Bi-Directional Join Relation

```graphql
type Parent{
    name: String
    child: Child @mongoRelation    # required! field which stores ids
}

type Child{
    name: String
    parent: Parent
}
```
Uni-Directional Join Relation

```graphql
type Parent{
    name: String
    child: Child    # optional. field which stores ids
}

type Child{
    name: String
 }
```

Embedded Type Relation

```graphql
type Parent{
    name: String
    child: Child      # optional. field that stores inline ids
}

type Child @embedded {
    name: String
}
```

Embedded Type To Other Non-Embedded Type Relation

```graphql
type Parent{
    name: String
    child: Child     # optional. field that stores inline ids
}

type Friend{
    name: String
}

type Child @embedded {
    name: String
    friend: Friend   # optional. field that stores inline ids
}
```

# Questions:

* always use name instead of field/column/name?
* discern between mongo/pg?
* relation tables in mongo

**Comment by sorenbs** [comment]
While I appreciate the passion displayed by the community in the comments here, this thread is a place to discuss design decisions for the upcoming MongoDB connector. I have deleted a number of comments that were off-topic. If you are interested in these other topics, feel free to reach out to me in slack or open a conversation in the forum.

**Comment by agolomoodysaada** [comment]
Take a look at [mongodb references](https://docs.mongodb.com/manual/reference/database-references/). Equivalent to foreign keys in pgsql.

It was also previously mentioned that multi-document transactions is supported by mongo 4.0.

**Comment by sorenbs** [comment]
As we are getting closer to the first beta release of the Mongo connector, we are looking for early adopters to take it for a spin.

If you are planning to build an application using Prisma and MongoDB, now is a good time to get involved to make sure the MongoDB connector covers all your needs.

If you are interested, please ping @do4gr here or in the Prisma slack 🙏

**Comment by agolomoodysaada** [comment]
@do4gr , I'm interested in testing it out... any docs I can follow?

**Comment by do4gr** [comment]
The documentation how to set up Prisma to use Mongo and how to use the new features can be found here: https://github.com/prisma/Mongo-Connector-Preview

**Comment by houmark** [comment]
Also interested in testing out the MongoDB connector.

**Comment by amilner42** [comment]
To those trying out this new connector, it would be great to leave links if your work is open source.

Hoping to try this myself soon.

**Comment by nhuesmann** [comment]
@do4gr I'd like to test the MongoDB connector and hope to use it in production on a project I'm currently building.

**Comment by do4gr** [comment]
The current documentation how to use it can be found here: https://github.com/prisma/Mongo-Connector-Preview

We'll release full documentation in the next days.

**Comment by schickling** [comment]
https://www.prisma.io/blog/mongodb-preview-ow4wahkekaep/ 🎉

**Comment by vibhanshu909** [comment]
I was waiting a long time for this, and I must say prisma is doing a great job, I was wondering if there is support for multi-document ACID transactions that were introduced in mongodb v4. I can see that the current documentation is using mongodb v3.6 not v4+. Thanks!

**Comment by do4gr** [comment]
Hey @vibhanshu909, 

we haven't done extensive testing on the compatibility with Mongo 4.0. The core is structured in a way that we should be able to test this with 4.0 fairly easily. The challenge is in properly testing this for correctness and performance afterwards. This is why we are still holding off on this. But I would expect us to enable this at some point.

**Comment by 001123** [comment]
Wow amazing, wish have support mongo 4.0 soon 👍

**Comment by pantharshit00** [comment]
You can track the progress on multi-document ACID transactions in #3785

**Comment by epileftro85** [comment]
Nothing about GeoJson? this future is mandatory for many MongoDB projects.
